test/unit/incoming-message-stream-test.js

```diff
@@ -175,4 +176,49 @@ describe('IncomingMessageStream', function() {
       incoming.resume();
     });
   });
+
+  it('should validate packet header size', function(done) {
+    const packetData = Buffer.from('test1234');
+    const packetHeader = Buffer.alloc(7);
+
+    let offset = 0;
+    offset = packetHeader.writeUInt8(0x11, offset);
+    offset = packetHeader.writeUInt8(0x01, offset);
+    offset = packetHeader.writeUInt16BE(8 + packetData.length, offset);
+    offset = packetHeader.writeUInt16BE(0x0000, offset);
+    packetHeader.writeUInt8(0x00, offset);
+
+    const packet = Buffer.concat([packetHeader, packetData]);
+
+    const incoming = new IncomingMessageStream(new Debug());
+
+    assert.throws(() => {
+      incoming.end(packet);
+    }, ConnectionError, 'Unable to process incoming packet');
+
+    done();
+  });
+
+  it('should validate packet buffer length', function(done) {
+    const packetData = Buffer.from('test1234');
+    const packetHeader = Buffer.alloc(8);
+
+    let offset = 0;
+    offset = packetHeader.writeUInt8(0x11, offset);
+    offset = packetHeader.writeUInt8(0x01, offset);
+    offset = packetHeader.writeUInt16BE(8 + 10, offset);
+    offset = packetHeader.writeUInt16BE(0x0000, offset);
+    offset = packetHeader.writeUInt8(1, offset);
+    packetHeader.writeUInt8(0x00, offset);
+
+    const packet = Buffer.concat([packetHeader, packetData]);
+
+    const incoming = new IncomingMessageStream(new Debug());
+
+    assert.throws(() => {
+      incoming.end(packet);
+    }, ConnectionError, 'Unable to process incoming packet');
+
+    done();
+  });
```

Comment:
I think this test here is actually showing the issue I mentioned over in https://github.com/tediousjs/tedious/pull/1102/files#r432504620.

The packet that is written here is fine. The packet says it's length is 18 bytes, but we're only writing 16 bytes of data here. In this case, the `IncomingMessageStream` should actually not emit an error - it should be waiting for more data coming in from the network instead.

This was working before, because of the `break` statement which would stop processing and wait for more data to come in.

Reply:
I'm just a bit confused. If the packet length specifies 18 bytes, wouldn't that mean that the current packet must be 18 bytes long? I'm assuming that the next data coming in is a new packet with a new packet header.

> 2.2.3.1.3 Length
> Length is the size of the packet including the 8 bytes in the packet header. It is the number of bytes
> from the start of this header _to the start of the next packet header._
>

Reply:
> If the packet length specifies 18 bytes, wouldn't that mean that the current packet must be 18 bytes long?

Yes, but a packet of 18 bytes could be delivered to the stream in any number of chunks. It could be two chunks being 9 bytes, or 4 chunks (1 byte, 2 bytes, 3 bytes, 13 bytes), or 18 chunks each containing 1 byte, or whatever other combination you could come up with. This depends on the underlying transport and the operating systems kernel. The only guarantee we have is that the data comes in in the same order as sent by the server, but there is no guarantee around the "chunkiness" of the arriving data.